### PR TITLE
DEV-2832 | delivery management for new transactional email recurring contribution reminder

### PR DESCRIPTION
**Pull request by benjaminEwhite**
Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

This is the first of three planned stacked PRs.

1. (https://github.com/newsrevenuehub/rev-engine/pull/835) (corresponding to [DEV-2832 | Delivery management for new transactional email: recurring contribution reminder](https://news-revenue-hub.atlassian.net/browse/DEV-2832) - **THE CURRENT PR**
2. (https://github.com/newsrevenuehub/rev-engine/pull/871) (corresponding to [DEV-2831 | Layout for new transactional email: recurring contribution reminder](https://news-revenue-hub.atlassian.net/browse/DEV-2831))
3. A future PR of branch DEV-708 (the base for this PR) into main. That PR has not yet been created in order to conserve review app slots. That branch will be solely for handling any merge conflicts ahead of merging into main.

#### What's this PR do?

This PR sets the groundwork for the parent ticket ([DEV-708 | Delivery management for new transactional email: recurring contribution reminder](https://news-revenue-hub.atlassian.net/browse/DEV-708)). That work is being tackled in two separate tickets/PRs — this one ([DEV-2832](https://news-revenue-hub.atlassian.net/browse/DEV-2832)), which is focused on triggering reminder emails and [DEV-2831](https://news-revenue-hub.atlassian.net/browse/DEV-2831) which will be focused on styling and finalizing content for the email template (and which @GuiMend  will be working on).

At a high level, this PR adds a webhook receiver for Stripe `invoice.upcoming` events. Stripe will send a webhook for subscriptions that have an upcoming invoice. [In the Stripe dashboard](https://dashboard.stripe.com/settings/billing/automatic), you can configure the number of days ahead of a new charge that this webhook gets sent. Stripe allows it to be set to 3, 7, 15, 30, or 45 days.

The new webhook receiever retrieves the contribution for the upcoming invoice (via the invoice's "subscription" value, which maps to `Contribution.provider_subscription_id` in the NRE domain model), and if the contribution's interval is YEARLY, calls a new model method (`Contribution.send_recurring_contribution_email_reminder`), which uses the pre-existing `send_templated_email` task to asynchronously send a reminder email the the contributor.

#### Why are we doing this? How does it help us?

Better user experience and in order to comply with regulations regarding payment reminders in the state of California.

#### How should this be manually tested? Please include detailed step-by-step instructions.

Manually testing this PR is rather involved, but possible. The manual testing strategy requires using Stripe's [test clock feature](https://stripe.com/docs/billing/testing/test-clocks) in order to trigger the `invoice.upcoming` webhook. For other Stripe webhooks, it's possible to trigger from the command line by running `stripe trigger <event-name>`, but `invoice.upcoming` is not presently supported.

Here's how to test:

1. Make an annual recurring contribution [here in the review app](https://billypenn-dev-2832.revengine-review.org/donate/). For the email address, use your NRH email address.
2. Go to the [Django admin](https://dev-2832.revengine-review.org/nrhadmin/contributions/contribution/), find your new contribution, and grab its ID. You'll need this momentarily.
3. Next go to [test-clocks section of the Stripe dashboard](https://dashboard.stripe.com/acct_1JhHQsPbhdfvD1kV/test/subscriptions) for the connected account that you contributed to in step 1 (note that it's crucial you do this from connected account, otherwise things won't work out). There, click the "new simulation" button. Add a unique name. You can go with the default clock time, and save.
4. You'll be taken to the detail view for your test clock. Click the "Add" button to add a customer. The values for the data you provide here aren't that important (the email doesn't actually need to match the email you used in the contribution).
5. After adding the customer, click the "Add" button a second time and choose the "Add subscription" option. In that interface, choose the customer you just created. Choose one of the $20 products. You can go with default values for everything else, except for Payment, where you should click "add payment method" and use the 424242.... fake cc info.
6. After adding the subscription, you'll be taken back to the clock detail view. Click on the subscription to get taken to its detail view and grab the subscription ID from the browser address bar (the URL will look something like `https://dashboard.stripe.com/test/subscriptions/sub_1M4plABMAMOLTEakrms2cOlo` — you want the `sub_...` value). Leave this browser tab open, as you'll need to come back to it in a moment.
7. Next, we need to add that subscription ID as the `provider_subscription_id` value for the contribution you created in step 8. To do that, you'll need to use the Heroku CLI to log in to the review app. From the command line, do ` heroku run bash --app rev-engine-dev-2832-del-u6muvc`. Then once you're logged in, do `python manage.py shell`. From there do the following:

    ```python
    from apps.contributions.models import Contribution
    contribution = Contribution.objects.get(id=<id from step 2>)   # substitute in value from step 2
    contribution.provider_subscription_id = "<subscription id from step 6>"  # substitute in value from step 6
    contribution.save()
    exit()
    ```

    You can now close this command line terminal or otherwise exit from the heroku bash shell.

8. Now back in the browser tab for the Stripe subscription from step 6, click the "Advance time" button. Click on the date input and then click the `+1 year` option. Wait for the test clock to finish processing.
9. Wait for up to ~5 minutes seconds or so for the webhook to process and the email to send, then check your inbox. You should see a reminder email containing data about the contribution (keeping in mind that this is not final copy or final layout). If the email hasn't arrived after 5 minutes, something probably went wrong with one of the previous steps, in which case you can reach out to @benjaminEwhite and he can dig into logs to determine what went wrong.
10. When you're done with all of the above, you can delete the test clock from Stripe.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

We should document for org users that payment reminders get sent ahead of contributions.

#### Have automated unit tests been added? If not, why?

Yes.

#### How should this change be communicated to end users?

Once we set the number of days ahead of next charge in Stripe dashboard and release this feature, we should communicate to existing orgs that contributors will receive reminder emails. 

#### Are there any smells or added technical debt to note?

No.

#### Has this been documented? If so, where?

No, but should be as per earlier comments.

#### What are the relevant tickets? Add a link to any relevant ones.

- [DEV-708 | Delivery management for new transactional email: recurring contribution reminder](https://news-revenue-hub.atlassian.net/browse/DEV-708)
- [DEV-2832 | Delivery management for new transactional email: recurring contribution reminder](https://news-revenue-hub.atlassian.net/browse/DEV-2832)
- [DEV-2831 | Layout for new transactional email: recurring contribution reminder](https://news-revenue-hub.atlassian.net/browse/DEV-2831) 

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

This is not an issue for review app, but after deploying to staging, test, and production, I believe we'll need to run the `create_stripe_webhook` management command in order for the `invoice.upcoming` webhook to start being sent in those environments.

Additionally, we'll need to set the `Upcoming renewal events
` [in the Stripe dashboard](https://dashboard.stripe.com/settings/billing/automatic) to the desired value (3, 7, 15, 30, or 45 days). 

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

[DEV-2831 | Layout for new transactional email: recurring contribution reminder](https://news-revenue-hub.atlassian.net/browse/DEV-2831)

### CODE REVIEW

File: apps/contributions/models.py [hunk 1]
```diff
@@ -345,3 +352,34 @@ def handle_thank_you_email(self, contribution_received_at=None):
...
+                # nb, we have to send this as pre-formatted because this data will be serialized
+                # when sent to the Celery worker
+                "charge_date": next_charge_date.strftime("%m/%d/%Y"),
+                "contribution_amount": self.formatted_amount,
+                "contribution_interval": self.interval,
+                "is_non_profit": self.donation_page.revenue_program.non_profit,
+                "contributor_email": self.contributor.email,
+                # todo -- update this after DEV-2519 merged
+                "tax_id": getattr(self.donation_page.revenue_program, "tax_id", "tax-id-coming-soon"),
+                "magic_link": (
```
Comment (njharman) [inline apps/contributions/models.py:380]: don't have to do the mark_safe() thing? like elsewhere with magic_link.

File: apps/contributions/webhooks.py [hunk 2]
```diff
@@ -140,3 +143,20 @@ def handle_subscription_canceled(self):
         contribution.payment_provider_data = self.obj_data
         contribution.status = ContributionStatus.CANCELED
         contribution.save()
+
+    def process_invoice(self):
+        if self.event.type == "invoice.upcoming":
+            self.process_upcoming_invoice()
+
+    def process_upcoming_invoice(self):
```
Comment (njharman) [inline apps/contributions/webhooks.py:151]: Why not just one method

```
def process_invoice...
    if not thing I want:
      return
    do the work.
```

Comment (njharman) [reply apps/contributions/webhooks.py:151]: I guess, maybe gonna be other types of invoices besides upcoming?

**Review by njharman — APPROVED** [review]
Couple minorish questions. Approved.

File: apps/contributions/models.py (hunk 1 above)
Comment (benjaminEwhite) [reply apps/contributions/models.py:380]: Yes, indeed!

Comment (benjaminEwhite) [reply apps/contributions/models.py]: b93d4afefb1

File: apps/contributions/webhooks.py (hunk 2 above)
Comment (benjaminEwhite) [reply apps/contributions/webhooks.py:151]: Yeah, I was on the fence and had similar thought. Decided to go with pattern that was already established in this file, but your question is enough to push me to going with single method.

Comment (benjaminEwhite) [reply apps/contributions/webhooks.py]: 50d6619ab1e

**Pull request merged**